TAP Migration

julian-londono opened the pull request:
This PR migrates all testing to [Node TAP](https://node-tap.org/).  In doing so, I've re-architected the project's testing to align better with best practices for testing of this sort.

Creates:
- `unit` tests: test function-level behavior for `pre-processors`
- `system` tests: test end-to-end behavior of the loader by spawning `Node` child processes and validating their output

~~Note:  Currently requires `loaderconfig.mjs` to be set up a certain way for tests~~
~~TODO~~: run tests regardless of `loaderconfig` file state --> Done!

jkrems commented on tap-snapshots/test-unit-ts.mjs-TAP.test.js:
```diff
@@ -0,0 +1,34 @@
+exports[`test/unit/ts.mjs TAP > Pre-Processor processes input correctly 2`] = `
+Object {
+  "extension": ".mjs",
+  "source": "import { dummy } from \\"/usr/local/google/home/jlondono/Documents/presm/test/fixtures/tsmodule1.ts\\";\\nconsole.log(dummy);\\n",
```
I assume this would fail on CI?

julian-londono replied:
Oops, yup, this should be changed to account for resolving specifiers to relative paths, not absolute.   This feature is coming to the ts-loader soon.

julian-londono replied:
Fixed this, it now fails for the known cause of replacing specifiers with absolute paths instead of relative one 👍

jkrems commented on test/fixtures/yamlTest.mjs:
```diff
@@ -0,0 +1,4 @@
+import yamlFile from './yamlExample.yaml';
+
+console.log('This is my YAML file import');
+console.log(yamlFile);
```
```suggestion
console.log(yamlFile);

```

jkrems commented on test/test-utils.mjs:
```diff
@@ -0,0 +1,103 @@
+export let cleanSnapshot = () =>
+  (tap.cleanSnapshot = snapshot => {
+    return snapshot.replace(
+      /(for file:\/\/\/.+?(?=\/presm))|((for \/.+?(?=\/presm)))/gm,
```
nit: I assume this works for the most part but I'm not sure it's all that reliable. Maybe you could build the URL of the project root and explicitly replace that?

julian-londono replied:
Makes sense, I'll try that!

julian-londono replied:
Done, makes more sense now, cleaner too!

jkrems commented on test/unit/ts.mjs:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+main();
```
I think this risks rejected promise warnings instead of failed tests. One common pattern is:

```js
tap.test('ts fixtures', async t => {
  await batchTest(t, /* ... */);
});
```

Right now some test failures may just turn into a `console.log` above. Bonus points: When iterating over test cases (in this case: fixture files), it can also be nice to explicitly create subtests (`t.test`) for each. For that you'd likely want to pass the tap interface to `batchTest`, e.g. as the first argument.

julian-londono replied:
Ah I see, I'll change to that pattern as well as explicitly calling subtests that way

julian-londono replied:
Changed this patter throughout, subtests do make more sense

jkrems commented on test/unit/yaml.mjs:
```diff
@@ -0,0 +1,23 @@
+  }
+}
+
+main();
```
See above for `async main` in tests.

julian-londono replied:
Same as above, will update this too

jkrems commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 *.log
 /build
 .vscode
+.nyc_output
```
```suggestion
.nyc_output

```

feywind commented on test/test-utils.mjs:
```diff
@@ -0,0 +1,103 @@
+ * @param inputs List of input source files to process
+ */
+export async function batchTest(processor, options, inputs) {
+  options = Array.isArray(options)
```
Style nit - I'd probably make a new variable here just so it's clear that the value may be different than the parameter farther down.

julian-londono replied:
Makes sense, I'll update

julian-londono replied:
Updated

feywind commented on test/test-utils.mjs:
```diff
@@ -0,0 +1,103 @@
+    ? options
+    : Array(inputs.length).fill(options);
+
+  let numTests = inputs.length;
```
Another style nit - not just here, but wherever, it's probably good to be in the habit of using `const` instead of `let` if the values won't change.

julian-londono replied:
Good tip, I'll update accordingly

julian-londono replied:
Fixed this everywhere

feywind commented on test/unit/ts.mjs:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+main();
```
I'm not super familiar with mjs, but is it possible to top-level await? Might want to do that or something to make sure errors get propagated out in a way that TAP will catch.

julian-londono replied:
Yup good point, I'll change this to TAP's own async test call, should allow TAP to catch errors that way

julian-londono replied:
Changed this pattern and tested for TAP catching errors, good call 👍

feywind commented on test/unit/yaml.mjs:
```diff
@@ -0,0 +1,23 @@
+  }
+}
+
+main();
```
Same comment as ts.mjs above.

julian-londono replied:
Same as above, will update

feywind commented on package.json:
```diff
@@ -26,8 +26,11 @@
-    "test-unit": "node --experimental-top-level-await ./tests/tests.mjs"
+    "test": "npm run test-unit && npm run test-system",
+    "test-snapshot": "npm run test-snapshot-unit && npm run test-snapshot-system",
+    "test-unit": "tap --no-coverage ./test/unit/",
```
I'm curious about why the `--no-coverage` - is it just that there aren't enough tests yet, and we want to get TAP working first?

julian-londono replied:
Allowing coverage triggers TAP's use of [`spawn-wrap`](https://www.npmjs.com/package/spawn-wrap).  This doesn't break the loader but adds more for the hooks to deal with imports so the snapshot gets messy.  Added `--no-coverage` to keep this test simple for now.  Good question though, I can explore coverage testing later.

julian-londono commented:
- Cleaned up the projects lint errors and warnings significantly
- Added the ability to test end-to-end (`system` tests) with a custom `loaderconfig.json` fixture for testing

I feel confident about the project's testing infrastructure to move forward in a test-driven manner from now on 👍
